doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+    match edit with
+    | [] -> prods
+    | Sedit "ADD_OPT" :: sym :: tl ->
+      let pfx = List.rev seen in
+      let posn = find_first edit0 prods nt in
+      let prods' = insert_after posn [pfx @ (Sopt sym :: tl)] prods in
+      let prods'' = remove_prod (pfx @ (sym :: tl)) prods' nt in
+      let prods''' = remove_prod (pfx @ tl) prods'' nt in
+      edit_single_prod_r tl prods''' nt seen
+    | Sedit "ADD_OPT" :: [] -> error "Bad position for ADD_OPT\n"; prods
+    | Sedit2 ("USE_NT", name) :: sym :: tl ->
+      let nt, _ = maybe_add_nt g nt name sym in
+      let pfx = List.rev seen in
+      let posn = find_first edit0 prods nt in
+      let prods' = insert_after posn [pfx @ (Snterm nt :: tl)] prods in
+      let prods'' = remove_prod (pfx @ (sym :: tl)) prods' nt in
+      edit_single_prod_r tl prods'' nt seen
+    | Sedit2 ("USE_NT", _) :: [] -> error "Bad position for USE_NT\n"; prods
+    | sym :: tl ->
+      edit_single_prod_r tl prods nt (sym :: seen)
+  in
+  edit_single_prod_r edit0 prods nt []
+
+let apply_edit_file g edits =
+  List.iter (fun b ->
+      let (nt, eprods) = b in
+      if not (edit_all_prods g nt eprods) then begin
+        let rec aux eprods prods add_nt =
+          match eprods with
+          | [] -> prods, add_nt
+          | (Snterm "DELETE" :: oprod) :: tl ->
+            aux tl (remove_prod oprod prods nt) add_nt
+          | (Snterm "DELETENT" :: _) :: tl ->  (* note this doesn't remove references *)
+            g.map <- NTMap.remove nt g.map;
+            g.order <- List.filter (fun nt1 -> nt1 <> nt) g.order;
+            aux tl prods false
+          | (Snterm "EDIT" :: oprod) :: tl ->
+            aux tl (edit_single_prod g oprod prods nt) add_nt
+          | (Snterm "REPLACE" :: oprod) :: (Snterm "WITH" :: rprod) :: tl ->
+            let posn = find_first oprod prods nt in
+            let prods' = match posn with
+            | None -> prods
+            | Some _ ->
+              let prods' = insert_after posn [rprod] prods in  (* insert new prod *)
+              remove_prod oprod prods' nt      (* remove orig prod *)
+            in
+            aux tl prods' add_nt
+          | (Snterm "REPLACE" :: _ as eprod) :: tl ->
+            error "Missing WITH after '%s' in '%s'\n" (prod_to_str eprod) nt;
+            aux tl prods add_nt
+          | prod :: tl ->
+            (* add a production *)
+            if has_match prod prods then
+              error "Duplicate production '%s' for %s\n" (prod_to_str prod) nt;
+            aux tl (prods @ [prod]) add_nt
+        in
+        let prods, add_nt =
+          aux eprods (try NTMap.find nt g.map with Not_found -> []) true in
+        if add_nt then begin
+          (* maybe add a new nt and its productions *)
+          if not (List.mem nt g.order) then
+            g.order <- g.order @ [nt];
+          g.map <- NTMap.add nt prods g.map
+        end
+      end)
+    edits
+
+
+(*** main routines ***)
+
+let nts_in_prod prod =
```

Comment:
It looks like you could return a set here rather than a list.

Reply:
This is called by nt_closure.  I want to preserve the order of the nonterminals.

Reply:
`nt_closure` is only called by functions that immediately turn its output to a set. There is no point in preserving an order you never use.

Reply:
Good point.  I was using the order for a long while.  I think it will be helpful when we're refining the order of nonterminals in the .mlg files (i.e., editing the code locally to get specific sets of productions).  I will add a todo saying to re-evaluate whether to sets later on.

Reply:
Github ate this comment:  I was using the nt_closure order for a long while.  I think it will be helpful as we review the .mlg grammars and figure out how to group them and what order we want them in (i.e. modifying the code locally to generate grammar subsets).  orderedGrammar is fairly random right now.  I added a todo on nt_closure to consider using sets later.